dice_ml/explainer_interfaces/dice_genetic.py

```diff
@@ -461,47 +371,64 @@ def find_counterfactuals(self, query_instance, desired_range, desired_class, fea
         iterations = 0
         previous_best_loss = -np.inf
         current_best_loss = np.inf
-        current_best_cf = []
         stop_cnt = 0
         cfs_preds = [np.inf]*self.total_CFs
-        while iterations < maxiterations:
-            if abs(previous_best_loss - current_best_loss) <= 1e-2: #and (self.model.model_type == 'classifier' and all(i == desired_class for i in cfs_preds) or (self.model.model_type == 'regressor' and all(desired_range[0] <= i <= desired_range[1] for i in cfs_preds))):
+        to_pred = None
+
+        while iterations < maxiterations or len(population) == self.total_CFs:
+            start = timeit.default_timer()
+            if abs(previous_best_loss - current_best_loss) <= thresh and (self.model.model_type == 'classifier' and all(i == desired_class for i in cfs_preds) or (self.model.model_type == 'regressor' and all(desired_range[0] <= i <= desired_range[1] for i in cfs_preds))):
                 stop_cnt += 1
             else:
                 stop_cnt = 0
             if stop_cnt >= 5:
                 break
             previous_best_loss = current_best_loss
-            population_fitness = []
-            current_best_loss = np.inf
-            current_best_cf = []
-            for k in range(self.population_size):
-                loss = self.compute_loss(population[k], desired_range, desired_class)
-                population_fitness.append((k, loss))
-
-                if loss < current_best_loss:
-                    current_best_loss = loss
-                    current_best_cf = population[k]
-
-            cfs_preds = [self.predict_fn(cfs) for cfs in current_best_cf]
-            # 10% of the next generation is fittest members of current generation
-            population_fitness = sorted(population_fitness, key=lambda x: x[1])
-            s = int((10 * self.population_size) / 100)
-            new_generation = [population[tup[0]] for tup in population_fitness[:s]]
-
-            # 90% of the next generation obtained from top 50% of fittest members of current generation
-            s = int((90 * self.population_size) / 100)
+
+            new_array = [tuple(row) for row in population]
```

Comment:
better not to mix python lists and np arrays. Since you convert to np array in the next line, can define `new_array` as a np.array too.

Another faster, pythonic option is to simply do `new_array=set(population)` and you will automatically get unique values. Then convert it to back to list or np.array, as `list(new_array)`